va/vampic.go

```diff
@@ -0,0 +1,320 @@
+package va
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"maps"
+	"math/rand/v2"
+	"slices"
+	"time"
+
+	"github.com/letsencrypt/boulder/core"
+	berrors "github.com/letsencrypt/boulder/errors"
+	bgrpc "github.com/letsencrypt/boulder/grpc"
+	"github.com/letsencrypt/boulder/identifier"
+	"github.com/letsencrypt/boulder/probs"
+	vapb "github.com/letsencrypt/boulder/va/proto"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	PrimaryPerspective = "primary"
+
+	challenge = "challenge"
+	caa       = "caa"
+	all       = "all"
+	pass      = "pass"
+	fail      = "fail"
+)
+
+// observeLatency records entries in the validationLatency histogram of the
+// latency to perform validations from the primary and remote VA perspectives.
+// The labels are:
+//   - operation: VA.ValidateChallenge or VA.CheckCAA as [challenge|caa]
+//   - perspective: [ValidationAuthorityImpl.perspective|all]
+//   - challenge_type: core.Challenge.Type
+//   - problem_type: probs.ProblemType
+//   - result: the result of the validation as [pass|fail]
+func (va *ValidationAuthorityImpl) observeLatency(op, perspective, challType, probType, result string, latency time.Duration) {
+	labels := prometheus.Labels{
+		"operation":      op,
+		"perspective":    perspective,
+		"challenge_type": challType,
+		"problem_type":   probType,
+		"result":         result,
+	}
+	va.metrics.validationLatency.With(labels).Observe(latency.Seconds())
+}
+
+// isPrimaryVA returns true if the VA is the primary validation perspective.
+func (va *ValidationAuthorityImpl) isPrimaryVA() bool {
+	return va.perspective == PrimaryPerspective
+}
+
+// mpicSummary contains multiple fields that are exported for logging purposes.
+// To initialize an empty mpicSummary, use newSummary(). The prepare a final
+// summary, use prepareSummary().
+type mpicSummary struct {
+	// Passed are the distinct perspectives that Passed validation.
+	Passed []string `json:"passedPerspectives"`
+
+	// Failed are the disctint perspectives that Failed validation.
+	Failed []string `json:"failedPerspectives"`
+
+	// RIRs are the distinct Regional Internet Registries that passing
+	// perspectives belonged to.
+	RIRs []string `json:"passedRIRs"`
+
+	// QuorumResult is the Multi-Perspective Issuance Corroboration quorum
+	// result, per BRs Section 5.4.1, Requirement 2.7 (i.e., "3/4" which should
+	// be interpreted as "Three (3) out of four (4) attempted Network
+	// Perspectives corroborated the determinations made by the Primary Network
+	// Perspective).
+	QuorumResult string `json:"quorumResult"`
+}
+
+// newSummary returns a new mpicSummary with empty slices to avoid "null" in the
+// JSON output. Fields are exported for logging purposes.
+func newSummary() mpicSummary {
+	return mpicSummary{[]string{}, []string{}, []string{}, ""}
```

Comment:
`summary` is declared and then captured by the `defer func()`, both here and in CheckCAA(). `newSummary()` tries to avoid the boilerplate of initializing it with empty slices in both spots. `prepareSummary()` is called by `remoteValidateChallenge()` but if that method is never executed because the local check resulted in failure then we don't want to log a zero-value `mpicSummary`. We want to log an empty summary to positively indicate that there were no remote validations attempted.